### PR TITLE
feat(web): Implement DOMException#code

### PR DESCRIPTION
**Pull request by yacinehmito**
Adds the `code` property to the class `DOMException`. It is pseudo-deprecated, as in new types of `DOMException` set it to zero. It is recommended for code to not rely on that value.
However, tests in the recently included web-platform-tests assert the value of the code, so we are forced to implement it. Issue #9009 highlights an example of such a test that rely on `code`.

Source of the values for each error type: https://developer.mozilla.org/en-US/docs/Web/API/DOMException#Error_names

cc @lucacasonato

### CODE REVIEW

**Comment by yacinehmito** [comment]
I could have done a map, but I wasn't sure whether it was going to pollute the global scope, so a `switch` it is.

**Comment by lucacasonato** [comment]
It doesn't pollute global scope. Could you switch to a `Map`? Should be less lines changed then. (and also would make it easy to implement the constants on `DOMException` in the future then).

File: op_crates/web/00_dom_exception.js
Comment (yacinehmito) [inline op_crates/web/00_dom_exception.js:29]: Nit: I don't like the redundancy of this with the `|| 0` part when setting `#code` in the constructor.
I'm tempted ro remove them both.

**Comment by kitsonk** [comment]
Not to be picky, but why use a map at all, why not just an object literal.  Creating any array of arrays to just populate a map seems arbitrarily complex.  🤷

**Comment by yacinehmito** [comment]
I can do it however you want it. I like `Map` because the `[property]` accessor grosses me out, but that's very particular of me.

I'll define an object literal.

**Comment by lucacasonato** [comment]
I don't like using a plain object in this case either, because what if somenone passes `toString` as the name to the DOMException constructor? `Map` prevents this. Consider `new Map(Object.entries({foo: 1}));`.

**Comment by yacinehmito** [comment]
I'll wait for @kitsonk's answer and I'll change it to what we'll settle on.

**Comment by 00ff0000red** [comment]
Just tossing it out there, but you could use a simple, plain, array literal, and `Array#indexOf`.

File: op_crates/web/00_dom_exception.js
Comment (00ff0000red) [inline op_crates/web/00_dom_exception.js:34]: Would the `??` op semantically fit better?

**Comment by yacinehmito** [comment]
@00ff0000red I'm not a fan of linear lookups.

File: op_crates/web/00_dom_exception.js
Comment (yacinehmito) [reply op_crates/web/00_dom_exception.js:34]: Indeed, the intent may be clear. Functionally though it is equivalent.

**Comment by 00ff0000red** [comment]
> @00ff0000red I'm not a fan of linear lookups.

There are no guarantees that object lookup, nor map lookups execute in constant time, I would presume that under the hood, V8 does a linear lookup for every suggestion here so far, with the exception of an object, as it may have to check the prototype chain, as pointed out by @lucacasonato, on top of that, V8 would have to check property descriptors, ex: checking whether or not the prop is a getter. As the map is a sorted map, you can't expect a better implementation than a linear search either.
But of course, the decision for the data structure is on you and the reviewers; the bigger problem with an array would probably be how hard/easy it would be to check which index is associated with which name, for example, someone skimming by to find the 12th code would have to read through the array until they've found the 12th one, whereas a Map would have the literal number 12 right beside it.

**Comment by yacinehmito** [comment]
There are also holes in the number. For example there are no error with code 2 or code 16. A `Map` will do just fine and is semantically clearer.

**Comment by yacinehmito** [comment]
@lucacasonato I pushed with your suggestion. I think we should merge; discussing more would be bike-shedding.

**Review by bnoordhuis — APPROVED** [review]
LGTM, thanks. The CI failure is because we're not at full WPT compliance yet, right?

Apropos Map vs Object: the prototype property issue can be avoided by using `Object.create()` or `{ __proto__: null, ... }`.

Plain objects have a minor advantage over maps in that they're not vulnerable to prototype pollution (think `Map.prototype.get = Map.prototype.toString`.)

**Comment by lucacasonato** [comment]
Test failure was because a WPT test was fixed in this PR, but had not been removed from the `expectFail` list. Corrected this now.

**Comment by yacinehmito** [comment]
> Plain objects have a minor advantage over maps in that they're not vulnerable to prototype pollution

This is a very good point.

I feel like we are down to only two options now (roughly):
- `Object.create(null)`, then set every mapping by assigning a new property.
- Using a `Map`, but storing the original implementation of `Map.prototype.get` then putting it in the map itself (or invoking it with `.call`).

I'm not export enough to figure out which one will maximise the performance of the lookup, but this is likely to be premature optimization.

@lucacasonato I defer to you to choose between the two options, and I'll implement it.

**Comment by kitsonk** [comment]
> `Object.create(null)`, then set every mapping by assigning a new property.

I would prefer this.

**Comment by yacinehmito** [comment]
Done, and pushed.

File: op_crates/web/00_dom_exception.js
Comment (00ff0000red) [inline op_crates/web/00_dom_exception.js:25]: There was a spelling error `nameToCodeMatting` => `nameToCodeMapping`
But this can be a single call anyway:
```suggestion
const nameToCodeMapping = Object.create(
  null,
  {
    IndexSizeError: { value: 1 },
    HierarchyRequestError: { value: 3 },
    WrongDocumentError: { value: 4 },
    InvalidCharacterError: { value: 5 },
    NoModificationAllowedError: { value: 7 },
    NotFoundError: { value: 8 },
    NotSupportedError: { value: 9 },
    InvalidStateError: { value: 11 },
    SyntaxError: { value: 12 },
    InvalidModificationError: { value: 13 },
    NamespaceError: { value: 14 },
    InvalidAccessError: { value: 15 },
    TypeMismatchError: { value: 17 },
    SecurityError: { value: 18 },
    NetworkError: { value: 19 },
    AbortError: { value: 20 },
    URLMismatchError: { value: 21 },
    QuotaExceededError: { value: 22 },
    TimeoutError: { value: 23 },
    InvalidNodeTypeError: { value: 24 },
    DataCloneError: { value: 25 },
  },
);
```

Comment (yacinehmito) [reply op_crates/web/00_dom_exception.js]: 1. Why wrap in an object with value?
2. Yes, one can also use `Object.assign`. There are many ways to do the same thing. However I proposed writing assignments and @kitsonk approved.

Comment (yacinehmito) [reply op_crates/web/00_dom_exception.js]: I didn't know about the second argument. Thanks.

**Comment by yacinehmito** [comment]
@lucacasonato I think this is mergeable now.

**Review by lucacasonato — APPROVED** [review]
Thanks

**Pull request merged**
